투표 API 연동

Na2te opened the pull request:
## 📌 연결된 이슈
- #69

### ✅ 작업 내용
- 투표 생성 API 작성
- 관련 `Request` 등 Model 생성
- `Studies DataSource` 생성 및 투표 생성 함수 및 의존성 등록
- `Repository`에 투표 생성 함수 작업 추가
- `UseCase` Provides 의존성 등록
- `ViewMode`l에서 투표 생성 API 관련 로직 추가
- `VoteScreen`에서 Scaffold 제거 및 공용 컴포넌트 TopAppBar 사용

### 📷 관련 이미지(영상)
<img width="300" alt="image" src="https://github.com/user-attachments/assets/38823e16-0ca7-4fb3-b8a6-fa9d6610b1d3" />

### 🤔 의논하고 싶은 내용(코드 리뷰 받고 싶은 부분)
- 아직 API 관련 백엔드에서 수정할 것들이 남아 있어서 데이터 클래스 수정할 예정입니다
- ~~SnackBar가 안나옵니다~~ => 해결했습니다
- getStudies 관련 함수가 깡통이라서 주석 처리했다가 일단 부분 TODO 처리했습니다

Na2te committed:
feat(#69): API 선언
- StudiesApi에 투표 생성 API 생성

Na2te committed:
feat(#69): Datasource 추가
- studiesApi 인자로 받고 Retrofit 처리할 StudiesDataSource 구현
- createVote API 작업 처리할 함수 구현

Na2te committed:
fix(#69): API 반환값 수정
- createVote API 통신으로 들어오는 data가 없기에 Unit으로 설정

Na2te committed:
feat(#69): 레포지토리 API 작업 구현
- createVote API 처리하는 함수 작업 완료
- domain에 인자로 받을 data class 생성
- datalayer의 CreateVoteRequest에 fromDomain() 확장함수 구현해서 domain에서 넘어온 인자로 Request 객체 생성

Na2te committed:
feat(#69): Usecase 추가
- CreateVoteUseCase 추가

Na2te committed:
feat(#69): DataSource 추가
- NetworkStudiesDataSource 의존성 binding

Na2te committed:
feat(#69): StudiesUseCaseModule 수정
- CreateVoteUseCase 의존성 binding

Na2te committed:
fix(#69): 데이터 타입 수정
- startTime과 endTime의 데이터를 String이 아니라 LocalDateTime으로 수정

Na2te committed:
feat(#69): TimeFormatter 수정
- VoteViewModel의 date와 time 값을 합쳐서 LocalDateTime을 만들어내는 함수 추가

Na2te committed:
feat(#69): VoteContract 수정
- Event를 sealed interface로 수정
- OnClickCompleteButton 이벤트를 studyGroupId라는 정수 인자 받는 data class로 수정
- ErrorContext 생성
- 투표 생성 작업 관련 에러 Context인 CreateVoteError 추가

Na2te committed:
fix(#69): MakeVote 탭 수정
- 인자로 studyGroupId를 받도록 수정

Na2te committed:
feat(#69): VoteViewModel 수정
- CreateVoteUseCase를 ViewModel 인자로 가져옴
- 완료 버튼(투표 생성 버튼) 눌렀을 때 해당 UseCase 실행하는 로직 생성
- handleException 구현

Na2te committed:
feat(#69): VoteScreen 수정
- Route에서 studyGroupId를 받도록 함
- 완료 버튼을 눌렀을 때, studyGroupId가 들어간 이벤트 발생
- TopAppBar에서 빠져있던 OnClick 눌렀을 때 함수 실행되도록 수정

Na2te committed:
fix(#69): VoteScreen 화면 수정
- Scaffold 벗기고 공용 컴포넌트의 TopAppBar 사용
- 안쓰는 TopAppBar 삭제

Na2te committed:
fix(#69): StudiesRepository, StudiesRepositoryImpl 수정
- getStudies가 깡통 API고, DataSource가 아니라 Api를 직접 참조하고 있어서 주석 처리를 했으나 TODO로 변경

Na2te committed:
Update MainScreen.kt

잃어버린 SnackBarHost를 찾아서 삼만리

Na2te committed:
fix(#69): 투표 생성 Request 데이터 수정
- 새로 추가된 속성 작업

Na2te committed:
fix(#69): CreateVote 수정
- 수정된 데이터에 따라 추가적인 속성 추가
- 종료 시간 설정 안할 경우 null 값 보내도록 수정

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/vote/VoteViewModel.kt:
```diff
@@ -50,9 +68,27 @@ constructor(
+                        showSuccessMessage("투표를 생성했습니다!")
+                        setEffect { VoteContract.Effect.NavigateToBackStack }
+                    }
+                }
```
최근에 든 생각이고 AI한테 조언받은건데 , 이런 방식으로 handle Event 함수가 굉장히 길어지더라구요 
저는 handleOnClickCompleteButton 같은 함수를 만들어서 이렇게라도 줄여야하나 싶은 생각이 있는데 영민님 생각은 어떠신지 궁금합니다

Na2te replied:
음 짧진 않지만 지금 이 코드로만 보면 사실 `VoteInfo` 객체를 만들 때 속성 하나에 한 줄을 써버리는 바람에 코드가 길게 된 것 같습니다

어 `ViewModel` 내/외부에 함수를 만들고 `is VoteContract.Event.OnClickCompleteButton` 분기 처리 내용을 해당 함수를 실행시켜서 간단하게 하자 라는 말씀이신 것으로 이해했는데,
하면 하는 거긴 합니다

다만 조금 생각할만한 건 앞서 말씀드렸다시피 일단 제 코드는 사실상 그냥 객체 생성 때문에 코드가 길어졌다는 점, 즉 코드가 딱히 중복되거나 로직이 복잡하다거나 긴 문제는 아닌 거 같습니다

보통 함수가 길어져서 다른 함수를 만드는 건 저는 크게 2가지로 생각하고 있습니다
다른 곳에서도 같은 로직을 사용하거나, 코드가 길어서 이해를 명확하게 하기 위해서

아마 본 코드의 경우 함수를 만들어 따로 빼게 된다면 후자의 경우가 될 것 같습니다

앞서 말씀드렸다시피 하면 하는 거긴 합니다만, 사실 그렇게 진행을 하게 되면, `handleEvent`의 분기에 따라 처리하는 다른 것들도 전부 내/외부에 함수로 빼서 사용하는 식으로 가야 하지 않을까 싶은 느낌도 있습니다
(한 쪽은 관련 내용을 함수로 뺀 다음 가져오고, 한 쪽은 그냥 로직 다 적어두고 그러면 코드 상으로 봤을 때 똑같은 분기 처리인데 이걸 뭐라고 해야 할지 모르겠지만 수준?이 조금 다르지 않나 싶습니다 추상화 수준이라고 하나요 알면 알려주세용)

여튼 그래서 결론은 해도 좋은데 적용하기 시작하면 다른 것들도 생각해야 해서 귀찮을 것 같다가 제 결론이긴 합니다
<p>$\normalsize{\color{#808080}제가\ 지금\ 결론\ 정리하고도\ 조금\ 어처구니\ 없긴\ 하네요...ㅋㅋ}$</p>

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/vote/VoteContract.kt:
```diff
@@ -1,27 +1,28 @@
-        data object OnDismissTimeButton : Event()
-        data class OnChangeDate(val date: Long) : Event()
-        data class OnChangeTime(val hour: Int, val minute: Int) : Event()
+    sealed interface Event : UiEvent {
```
`sealed class`에서 `sealed interface`으로 변경하는 이유가 궁금합니다!

Na2te replied:
> `sealed class`에서 `sealed interface`으로 변경하는 이유가 궁금합니다!

단순하게 `Event` 클래스 자체에서 가지는 변수가 아무것도 없기 때문입니다!
클래스로 하면 매번 상속 받을 때마다 `:Event()` 꼴로 해야 하는데 인터페이스면 `:Event` 만 해도 되니까요

todayis-sunny reviewed (approved):
커밋 단위가 되게 잘 나누어져 있네요!
작업하시느라 고생하셨습니다:)

Na2te committed:
fix(#69): 투표 생성 API 주소 수정

Na2te committed:
fix(#69): 투표 생성 Request Model 수정
- choose라는 선택 항목 추가 허용 옵션 추가

Na2te committed:
fix(#69): 투표 생성 Domain Model 수정
- choose라는 선택 항목 추가 허용 옵션 추가

Na2te committed:
fix(#69): 투표 생성 ViewModel 수정
- api 호출 시 선택 항목 추가 허용 항목 추가
- groupId를 Screen단에서 클릭 했을 때 넘겨받지 않고, Navigation에서 saveStateHandle로 들어간 인자를 꺼나서 씀
- 투표 생성 탭에서 인자를 받으므로 !!로 null이 아님을 처리

Na2te committed:
fix(#69): 종료 30분 전 알림 로직, groupId 전달 방식 수정
- 기존에는 종료 시간 설정 상관 없이 선택이 가능했지만 이제 종료 시간을 선택 안하면 비활성화 처리
- 기존에 Screen에서 생성 버튼을 눌렀을 때 Screen이 가지고 있던 그룹 번호를 람다 함수로 넘겼음
- 이제는 ViewModel에서 직접 savestatehandle에서 꺼내서 사용하므로 이벤트 등으로 인자를 건네주던 것 삭제

Na2te committed:
Merge branch 'develop' into feat/#69-voteapi
